### PR TITLE
Implementations of AES and GCM for RISC-V

### PR DESCRIPTION
**Pull request by hbrausen**
The included commits incrementally add the following functionality:

* Support for ASM implementations on the linux64-riscv64 target.
* Four-table implementation of AES in generic 64-bit RISC-V assembly (uses basic integer instructions only).
* Stub implementation of RISC-V CPUID and support for OPENSSL_riscvcap environment variable.
* Carryless multiplication (clmul) accelerated Galois field multiplication for GCM using RISC-V Zbb,Zbc extensions.
* Fast byte-swapping routines for RISC-V with Zbb extension.

An implementation of AES is provided in RISC-V assembly,
using only the RV64I base integer instruction set.

RISC-V has ratified additional bit-manipulation instructions in the
Zb[abcs] extensions, which include instructions for byte-reverse,
rotation, and carryless-multiplication. This series uses these
instructions to provide optimized performance for the GCM mode
on RV64 cores that implement Zb[abcs].

This series also introduces a basic implementation of RISC-V CPUID
and support for enabling ISA extension support at runtime using the
OPENSSL_riscvcap environment variable.

# Testing
For testing, the following has been done:

- `demos/cipher/aesgcm.c` was built and run under qemu, with both `OPENSSL_riscvcap="rv64gc_zba_zbb_zbc_zbs"` and `OPENSSL_riscvcap=""`. In both cases, the GCM tag was verified as correct.

- Built-in test suite was run using user-mode qemu. All tests pass. A few different variations of testing were run. Details are below:

<details><summary>Run of built-in tests (Click to Expand)</summary>

First, configuring the build (and disabling unrelated feature afalgeng as it does not run with my setup):

```
$ CC="riscv64-unknown-linux-gnu-gcc" ./Configure linux64-riscv64 no-afalgeng
$ make -j <num threads>
```

Next, I ran the built-in test suite using `make test`. The summary is below:

```
All tests successful.
Files=243, Tests=3333, 1420 wallclock secs ( 6.63 usr  0.51 sys + 1375.74 cusr 41.87 csys = 1424.75 CPU)
Result: PASS
```

Next, I enabled the Zba,Zbb,Zbc and Zbs extensions on my user-mode qemu, and ran the test suite again, this time with the environment variable OPENSSL_riscvcap defined:

```
$ export OPENSSL_riscvcap="rv64gc_zba_zbb_zbc_zbs"
$ export QEMU_CPU="any,x-zba=true,x-zbb=true,x-zbc=true,x-zbs=true"
$ make test
```

This gave the same result -- all tests passed.

Finally, I recompiled specifically for a machine with zba,zbb,zbc,zbs extensions:

```
$ CC="riscv64-unknown-linux-gnu-gcc -march=rv64gc_zba_zbb_zbc_zbs" ./Configure linux64-riscv64 no-afalgeng
$ make -j <num threads>
$ export OPENSSL_riscvcap="rv64gc_zba_zbb_zbc_zbs"
$ export QEMU_CPU="any,x-zba=true,x-zbb=true,x-zbc=true,x-zbs=true"
$ make test
```

Once again, all tests passed.

</details>

### CODE REVIEW

File: crypto/modes/asm/ghash-riscv64.pl
Comment (FdaSilvaYY) [inline crypto/modes/asm/ghash-riscv64.pl:49]: typo: extention

File: crypto/riscvcap.c
Comment (FdaSilvaYY) [inline crypto/riscvcap.c:77]: Style nit: please test pointers explicitly against NULL 
`    if (trigger != NULL)`

File: crypto/riscv64cpuid.pl
Comment (FdaSilvaYY) [inline crypto/riscv64cpuid.pl:75]: ASM Newbie question : `4:` is a jump label, no ?
Or is it something else that can be repeated ?

File: doc/man7/openssl-env.pod
Comment (FdaSilvaYY) [inline doc/man7/openssl-env.pod:77]: nit : there is a Copyright downside (line 94) who could be updated too.

**Review by FdaSilvaYY — COMMENTED** [review]
Very quick reading.
These changes will require CLA : 
https://www.openssl.org/policies/cla.html

File: crypto/riscv64cpuid.pl
Comment (hbrausen) [reply crypto/riscv64cpuid.pl:75]: `4:` is a numbered local label. (See section "Local Labels" at https://sourceware.org/binutils/docs/as/Symbol-Names.html)

There is no restriction on reuse of such labels. These labels are referred to by either writing `Nf` for forward reference, or `Nb` for backward reference. The assembler will jump to the next or previous matching numbered label, respectively.

**Comment by hbrausen** [comment]
@FdaSilvaYY Thank you for your comments.
I have made the suggested changes.

Also, FYI, the relevant ICLA and CCLA have been submitted.

**Comment by paulidale** [comment]
I didn't realise that qemu supported multiple versions of the RISC-V architecture.
It would be nice if our unit tests covered more than they currently do -- not here but in another pull request.

The `.github/workflows/cross-compiles` action is probably the relevant file for this.

**Comment by hbrausen** [comment]
I can see that one check failed, due to missing function prototypes. I suppose I must supply prototypes for the functions mentioned in the warnings in order to suppress them. Details:

<details>
  <summary>Click to expand</summary>

```
Run make -s -j4
In file included from crypto/riscvcap.c:18:
include/crypto/riscv_arch.h:29:23: error: function declaration isn't a prototype [-Werror=strict-prototypes]
   29 |     static inline int RISCV_HAS_##NAME()                            \
      |                       ^~~~~~~~~~
include/crypto/riscv_arch.def:21:1: note: in expansion of macro 'RISCV_DEFINE_CAP'
   21 | RISCV_DEFINE_CAP(ZBA, 0, 0)
      | ^~~~~~~~~~~~~~~~
include/crypto/riscv_arch.h:29:23: error: function declaration isn't a prototype [-Werror=strict-prototypes]
   29 |     static inline int RISCV_HAS_##NAME()                            \
      |                       ^~~~~~~~~~
include/crypto/riscv_arch.def:22:1: note: in expansion of macro 'RISCV_DEFINE_CAP'
   22 | RISCV_DEFINE_CAP(ZBB, 0, 1)
      | ^~~~~~~~~~~~~~~~
include/crypto/riscv_arch.h:29:23: error: function declaration isn't a prototype [-Werror=strict-prototypes]
   29 |     static inline int RISCV_HAS_##NAME()                            \
      |                       ^~~~~~~~~~
include/crypto/riscv_arch.def:23:1: note: in expansion of macro 'RISCV_DEFINE_CAP'
   23 | RISCV_DEFINE_CAP(ZBC, 0, 2)
      | ^~~~~~~~~~~~~~~~
include/crypto/riscv_arch.h:29:23: error: function declaration isn't a prototype [-Werror=strict-prototypes]
   29 |     static inline int RISCV_HAS_##NAME()                            \
      |                       ^~~~~~~~~~
include/crypto/riscv_arch.def:24:1: note: in expansion of macro 'RISCV_DEFINE_CAP'
   24 | RISCV_DEFINE_CAP(ZBS, 0, 3)
      | ^~~~~~~~~~~~~~~~
crypto/riscvcap.c:35:6: error: no previous prototype for 'strtoupper' [-Werror=missing-prototypes]
   35 | void strtoupper(char *str)
      |      ^~~~~~~~~~
crypto/riscvcap.c:49:6: error: no previous prototype for 'parse_env' [-Werror=missing-prototypes]
   49 | void parse_env(const char *envstr)
      |      ^~~~~~~~~
cc1: all warnings being treated as errors
make[1]: *** [Makefile:16698: crypto/libcrypto-lib-riscvcap.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [Makefile:3377: build_sw] Error 2
Error: Process completed with exit code 2.
```

</details>

I can also anticipate another potential build issue: Lines 57 and 58 in `ghash-riscv64.pl`:

```
# Enable zbb (basic bit manipulation) and zbc (carryless multiplication)
.option arch,+zbb,+zbc
```

The `.option arch,+xxx` directive for RISC-V was added to binutils very recently, so I'm not sure if it will be supported by the automated build.

I will push an update that adds the requested prototypes, and build with `-Werror=strict-prototypes` to test before pushing.

**Comment by t8m** [comment]
The CI failure is relevant.

**Comment by hbrausen** [comment]
Fixed warnings and built using -Werror, --strict-warnings Configure option to check.

**Review by pkubaj — CHANGES_REQUESTED** [review]
Could you also modify BSD-riscv64 target to use the same asm_arch as linux64-riscv64?

**Comment by hbrausen** [comment]
Minor update:

- Rebase on `master`
- Move an `add` instruction earlier in the inner loop of `AES_[en/de]crypt` to help with instruction fusion on platforms that can fuse `add rd,ra,rb; ld rc,0(rd)` sequences.

**Comment by hbrausen** [comment]
> Could you also modify BSD-riscv64 target to use the same asm_arch as linux64-riscv64?

We'll need to check that the asm we're providing is compatible with the llvm toolchain, and make any changes required.
We're investigating this now :)

**Comment by hbrausen** [comment]
> > Could you also modify BSD-riscv64 target to use the same asm_arch as linux64-riscv64?
> 
> We'll need to check that the asm we're providing is compatible with the llvm toolchain, and make any changes required. We're investigating this now :)

LLVM does not currently implement the `.option arch` directive that we are using to enable ISA extension support in our asm, though there is discussion [here](https://github.com/riscv-non-isa/riscv-asm-manual/pull/67) about making it standard. It looks like it will likely be supported in the near future, though enabling it right now would break the build on BSD-riscv64.

If we want to enable the `asm_arch` for BSD riscv64 right now, we would need some other way to tell the build to assemble this file with support for zbb,zbc extensions. Maybe some temporary workaround? Or else we delay enabling asm_arch for BSD until the assembler supports this directive. (Most of the gains from asm come from making use of ISA extensions, so this could be reasonable...)

**Comment by pkubaj** [comment]
OK, I guess it's most reasonable to wait. Thanks for explanation!

**Comment by kroeckx** [comment]
My understanding of ".option arch" is that it's a runtime selection by
the dynamic linker to select one of the implementations. What most of
our other assembler support does is move the runtime detection to
what the CPU supports and what is the best available implementation to
the C code in OpenSSL itself. This has the added benefit that you can
be more flexible in which implementation you pick, the presence of
some instructions isn't always a good indicator of the best performace.

**Comment by ptomsich** [comment]
> My understanding of ".option arch" is that it's a runtime selection by the dynamic linker to select one of the implementations. What most of our other assembler support does is move the runtime detection to what the CPU supports and what is the best available implementation to the C code in OpenSSL itself. This has the added benefit that you can be more flexible in which implementation you pick, the presence of some instructions isn't always a good indicator of the best performance.

".option arch" only instructs the assembler to modify the select "-march=" setting and does not otherwise interact with the runtime selection by the linker (ifunc would do this, which adds some additional machinery and directives … search for "@gnu_indirect_function" to see examples).  However, options like ".option arch" (on AArch64, the equivalent directives are ".arch" and ".arch_extension", see https://sourceware.org/binutils/docs/as/AArch64-Directives.html for details) it is typically used together with runtime-resolution of the best implementation (whether ifunc or a project-specific solution using function pointers) to override the default "-march=" setting and enable assembly of otherwise unsupported instructions.

**Comment by kroeckx** [comment]
So it just enables the assembler to understand instructions?

We've had to do encoding by hand (.byte?) for other cases where not
all supported assemblers understand an instruction.

**Comment by hbrausen** [comment]
@kroeckx Yes, that should be correct. Given that, hand encoding could be an interim solution while we wait for `.option arch` support to propagate through the ecosystem.

Note about CI build binutils:

CI builds with ubuntu-latest. The latest binutils is 2.38, which includes `.option arch`. Given the release schedule of ubuntu, there is a chance binutils 2.38 makes it into the next release, which would make this feature available in CI on Apr 21st 2022 at the latest.

**Comment by hbrausen** [comment]
Ubuntu released version 22.04 LTS yesterday, which supports `.option arch` in its cross binutils. With this release, CI should be able to build this pull request.

To test this configuration, I installed Ubuntu 22.04 LTS in a virtual machine, and verified that the riscv64 cross build completed properly. I was able to run the test suite under QEMU, in the manner described in `cross-compiles.yml` in the CI config.

When the `linux64-riscv64` target is built and `make test` is run, bitmanip is disabled by default. To test with bitmanip enabled, set the following additional environment variables:

```
OPENSSL_riscvcap="rv64gc_zbb_zbc"
QEMU_CPU="any,zba=true,zbb=true,zbc=true,zbs=true"
```

Towards enabling CI tests with zbb, zbc enabled, there is one complication: `qemu 6.2.0`, included in Ubuntu 22.04 LTS, lists the bitmanip extensions as "experimental", requiring the following syntax:

`QEMU_CPU="any,x-zba=true,x-zbb=true,x-zbc=true,x-zbs=true"`

Newer versions of QEMU remove the `x-` prefix, as these features are no longer experimental. Because of this change, if we write a new CI test to exercise bitmanip, we need to detect the version of QEMU present, and set `QEMU_CPU` accordingly. 

We may wish to wait for non-experimental support for bitmanip in the released QEMU before writing a dedicated CI test for bitmanip.

**Comment by cmuellner** [comment]
@t8m: can you please re-start the CI tests?

What else needs to be done to get this merged?

**Comment by hbrausen** [comment]
I had mistakenly assumed that `ubuntu-latest` in the context of GitHub CI referred to the latest official release of Ubuntu. This is not the case -- GitHub manually updates `ubuntu-latest` to point to the latest ubuntu release, with a significant delay. (With Ubuntu 20.04 LTS, this additional delay was about 6 months.) As a result, the binutils in GitHub CI still does not understand the `.option arch` directive, and CI fails.

**Comment by cmuellner** [comment]
If we need a green light from the CI before this can get merged, then we have a dependency to https://github.com/actions/virtual-environments/issues/5428.

**Comment by paulidale** [comment]
OpenSSL generally supports a wide range of tool chain versions.  I think it would be better to avoid `.option arch` completely even if this means hand coding instructions.

**Comment by t8m** [comment]
You should be able to use the perl translation machinery for that.

**Comment by hbrausen** [comment]
> You should be able to use the perl translation machinery for that.

I will modify the patchset to assemble the bitmanip insns in-place in perl, to relax dependency on toolchain version.

**Comment by hbrausen** [comment]
> OpenSSL generally supports a wide range of tool chain versions. I think it would be better to avoid `.option arch` completely even if this means hand coding instructions.

This is now complete: ghash-riscv64.pl assembles the instructions `clmul`, `clmulh`, and `rev8` directly into machine code, eliminating any dependency on `.option arch`.

I have tested the latest version as follows:

- Ran the built-in test suite under QEMU, testing with extensions both enabled, and disabled (via `OPENSSL_riscvcap`).
- Ran the demo `demos/cipher/aesgcm`, which generated the correct GCM tag.
- Compiled on Ubuntu 20.02 using binutils 2.34

@paulidale Could we re-trigger CI for this PR please?

File: crypto/modes/gcm128.c
Comment (ZenithalHourlyRate) [inline crypto/modes/gcm128.c:31]: `__riscv_xlen == 64` should also be checked here IMHO?

File: crypto/modes/gcm128.c
Comment (ZenithalHourlyRate) [inline crypto/modes/gcm128.c:705]: About the function name, I wonder whether the asm snippet needs the whole `rv64gc` to work, I think `rv64i` is sufficient?

Comment (hbrausen) [reply crypto/modes/gcm128.c:705]: I agree. I will change the name to reference `rv64i`.

File: crypto/modes/gcm128.c
Comment (hbrausen) [reply crypto/modes/gcm128.c:700]: You are correct, I will make this change. Thank you for catching this!

File: crypto/modes/gcm128.c
Comment (hbrausen) [reply crypto/modes/gcm128.c:31]: I do not agree with this change, but the reason is a bit complicated:

The code, with additional context, is:

```
/* RISC-V uses C implementation of gmult as a fallback. */
#if defined(__riscv)
# define INCLUDE_C_GMULT_4BIT
#endif
```

I do not think we should check for `XLEN == 64` here, because a future 32-bit RISC-V implementation of GCM may also want to use the C impl of `gcm_gmult_4bit` as a fallback.

(Other platforms provide a generic (using few/no extensions) fallback ghash or gmult routine in asm, but until someone has a need for a hand-optimized generic rv64i gmult, it is better to use the well-tested C routine as a fallback. Such a routine can always be contributed by someone later on.)

IMO, the correct place to check for `XLEN==64` is before enabling the `rv64i_zbb_zbc` impl specifically.

Comment (ZenithalHourlyRate) [reply crypto/modes/gcm128.c:31]: I'm convinced, though I think adding this check for integrity now and removing the check when there is a "future 32-bit RISC-V implementation of GCM" is also acceptable.

**Comment by hbrausen** [comment]
- Rebased on `master`
- Small changes to naming of asm routines for galois field multiplication
- Add explicit check for `XLEN==64` before enabling 64-bit galois field multiplication routines.

The following tests were run:
- Built-in test suite without defining OPENSSL_riscvcap (default)
- Built in test suite with OPENSSL_riscvcap="rv64gc_zba_zbb_zbc_zbs"
- Run of demos/cipher/aesgcm without defining OPENSSL_riscvcap (default) 
- Run of demos/cipher/aesgcm with OPENSSL_riscvcap="rv64gc_zba_zbb_zbc_zbs"

CI passed on our previous patch set. As these changes are minor, these changes should also pass.

@paulidale May we re-trigger CI?

**Comment by ZenithalHourlyRate** [comment]
FYI I've made a benchmark of this PR against the C baseline at https://gist.github.com/ZenithalHourlyRate/7b5175734f87acb73d0bbc53391d7140, I think you might be interested.

**Comment by hbrausen** [comment]
Hi folks,

Is there anything more that must be done to have this merged? Thank you!

File: crypto/riscvcap.c
Comment (paulidale) [inline crypto/riscvcap.c:57]: `OPENSSL_strlcpy` would be more sensible here.  This is a security toolkit and lots of eyes look for overflows.

File: crypto/riscvcap.c
Comment (paulidale) [inline crypto/riscvcap.c:62]: This should be `BIO_snprintf`

File: crypto/riscvcap.c
Comment (hbrausen) [reply crypto/riscvcap.c]: This change has been pushed, thank you!

File: crypto/riscvcap.c
Comment (hbrausen) [reply crypto/riscvcap.c]: This change has been pushed, thank you!

**Comment by hbrausen** [comment]
- Rebased on master
- Use `OPENSSL_strlcpy()` and `BIO_snprintf()` in riscv capability string parser.

The following tests were run:

- Built-in test suite
- Run of demos/cipher/aesgcm without defining OPENSSL_riscvcap (default)
- Run of demos/cipher/aesgcm with OPENSSL_riscvcap="rv64gc_zba_zbb_zbc_zbs"

CI passed on our previous patch set. As these changes are minor, these changes should also pass.

May we retrigger CI for the new changes?

File: crypto/riscvcap.c
Comment (paulidale) [inline crypto/riscvcap.c:40]: We don't normally put {} around a single statement body.

**Review by paulidale — APPROVED** [review]

File: crypto/aes/asm/aes-riscv64.pl
Comment (ZenithalHourlyRate) [inline crypto/aes/asm/aes-riscv64.pl:960]: The comment here is incorrect (found this when editing my PR)

Comment (hbrausen) [reply crypto/aes/asm/aes-riscv64.pl:960]: Thank you for catching this -- stale comments are always frustrating!

File: crypto/riscvcap.c
Comment (hbrausen) [reply crypto/riscvcap.c]: Fixed in latest push. Thank you!

**Comment by hbrausen** [comment]
- Rebased on master
- Updated stale comment in `aes-riscv64.pl`
- Fixed bracing to match code style in `riscvcap.c`

File: crypto/aes/asm/aes-riscv64.pl
Comment (ZenithalHourlyRate) [reply crypto/aes/asm/aes-riscv64.pl:960]: You fixed it in `Register assignment for AES_set_[en/de]crypt_key` but the comment for `Register assignment for AES_set_decrypt_key` is still stale.

BTW, `Register assignment for AES_set_[en/de]crypt_key` should be `Register assignment for AES_set_encrypt_key` as you have `clear_regs` in line 953. Thus these reg assignments in 670 do not cover `decrypt`

Comment (hbrausen) [reply crypto/aes/asm/aes-riscv64.pl]: Fixed in latest push, thank you!

**Comment by hbrausen** [comment]
- Rebased on master
- Updated stale comment in `aes-riscv64.pl`

We need one more review by a committer to approve these changes.
May we also retrigger CI? Thank you!

**Comment by cmuellner** [comment]
I'd like to note that this PR has been successfully tested by @kuroro99 as reported here: https://github.com/openssl/openssl/pull/18197#issuecomment-1126867231

**Review by t8m — APPROVED** [review]
@paulidale still OK after the rebase?

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks for the contribution.

**Pull request closed without merging**
